README.md
Emzee_Dev
=========
Provides a handy object inspection method for Magento developers. 

Facts
-----
- version: 1.0.0
- extension key: Emzee_Dev
- [extension on GitHub](https://github.com/mzeis/Emzee_Dev)

Usage
-----
To be honest, I had forgotten there is any other method besides info() in Emzee_Dev.
I guess that most of the other functionalities is covered by [FireGento_Debug](https://github.com/firegento/firegento-debug) and others now.
I'll list them here for the sake of completeness.

### info()

The most useful function is `info()`. You can call it using the model or the helper:

    # using a singleton model
    echo Mage::getSingleton('emzee_dev/debug')->info($variable);
    
    # using a helper
    echo Mage::helper('emzee_dev')->info($variable);
    
`info()` is intended to be used with objects. Providing info() with other types will
get you a simple `print_r()` output.

For objects, you get

* A **graphical presentation** of the public methods that exits in the class and its parent
  classes.
  - You can see which class defines the method.
  - Hover over the method name to see the defining class.
  - Click on the method name to see the method declaration (variables, default values).
* The names of the parent and child blocks if the object is a **block class**.
* A nice **backtrace** showing the content of the files next to the called line.

Here is a screenshot so you get a better picture:

![Emzee_Dev info() call](https://github.com/mzeis/Emzee_Dev/raw/master/app/code/community/Emzee/Dev/documentation/screenshot-info.png)

### configInfo()

* Prints the absolute pathes of various Magento folders.
* Prints a list of all modules with status, code pool, dependencies and version.

### memoryUsage()

Returns current memory usage in human readable format.

### modelInfo()

Returns information on models:

* which model groups are defined in config XML files (node global > models)
* what does that model group get translated to (e.g. "admin" => "Mage_Admin_Model")
* if it is a "normal" model: what is the resource model for this model
* if it is a resource model: which entities do exist for this model

Compatibility
-------------
- Magento CE 1.7 - 1.9 (only tested here; initial version worked with 1.3 CE, so you may try it with older versions)

Installation Instructions
-------------------------
1. Install the extension using [modman](https://github.com/colinmollenhour/modman) or copy all the files into your
   document root. If you copy the files yourself copy `app/design/frontend/base/default/template/emzee_dev/` to
   `app/design/adminhtml/default/default/template/emzee_dev/` manually.

Uninstallation
--------------
1. Remove the extension like all other extensions you install using modman.

Changelog
---------

See [Changelog](https://github.com/mzeis/Emzee_Dev/blob/master/CHANGELOG.md).

Support
-------
If you have any issues with this extension, open an issue on [GitHub](https://github.com/mzeis/Emzee_Dev/issues).

Contribution
------------
Any contribution is highly appreciated. The best way to contribute code is to open a [pull request on GitHub](https://help.github.com/articles/using-pull-requests).

Developer
---------
Matthias Zeis
[http://www.matthias-zeis.com](http://www.matthias-zeis.com)  
[@mzeis](https://twitter.com/mzeis)

Licence
-------
[OSL - Open Software Licence 3.0](http://opensource.org/licenses/osl-3.0.php)

Copyright
---------
(c) 2014 Matthias Zeis


CHANGELOG.md
Changelog
=========

1.0.0 (31.08.2014)
-----
* Added modman symlink so Emzee_Dev can be used in backend too

0.1.1 (16.06.2013)
-----
* Closed #7 - call to info() doesn't work inside upgrade script

0.1.0 (20.05.2013)
-----
* Closed #1 - Use Developer Client Restriction IPs for check
* Closed #2 - Remove mark / benchmark functionality
* Closed #3 - Remove static methods
* Closed #4 - Move view code to view layer
* Closed #5 - Make extension translatable
* Closed #6 - Make installable via Composer

0.0.1 (18.05.2013)
-----
* Initial version


app/code/community/Emzee/Dev/Block/Info.php
<?php

class Emzee_Dev_Block_Info extends Mage_Core_Block_Template
{
    /**
     * Returns the CSS code for the debug container.
     * The element id is used to raise the CSS specificity of the rules.
     * 
     * @param  string $id Identifier for the container
     * @return string
     */
    protected function _getCssCode($id)
    {
        return "<style type='text/css'>
        .emzee-dev-debug-container {
            background-color:   #fff;
            border:             1px solid #ddd;
            padding:            0;
            position:           absolute;
            text-align:         left;
            width:              960px;
            z-index:            10000;
        }
        
        #$id h1.handle {
            background-color:   #e6e6e6;
            border-bottom:      1px solid #999;
            margin:             0;
            padding:            3px;
        }
        
        #$id span#emzee-dev-debug-main-toggle {
            background-color:   #686868;
            color:              #fff;
            float:              right;
            font-size:          16px;
            padding:            5px;
        }
        
        #$id .handle:hover {
            cursor:             move;
        }
        
        #$id h2 {
            height:             14px;
            line-height:        17px;
            margin:             16px 0 8px;
            padding:            0 5px;
        }
        
        #$id ul {
            /*list-style-type:  disc;*/
            margin-left:        15px;
            padding:            0 3px;
        }
        
        #$id .class-color {
            border-left:        14px solid #fff;
            padding-left:       5px;
        }
        
        #$id .expandable:hover {
            cursor:             pointer;
        }
        
        #$id table {
            border:             1px solid #ddd;
            border-width:       1px 0;
            border-collapse:    collapse;
            width:              100%;
        }
        
        #$id table#backtrace-list div.filename {
            background-color:   #cccccf;
            font-weight:        bold;
            padding:            2px 5px;
        }
        
        #$id th, #$id td {
            border:             1px solid #ddd;
            padding:            4px 6px;
        }
        
        #$id th {
            background-color:   #eee;
        }
        
        #$id th.index, #$id td.index {
            text-align:         center;
        }
        
        #$id td.backtrace-info {
            padding:            0;
        }
        
        #$id ol.codeExcerpt {
            border:             0 solid #ddd;
            border-width:       1px 0 0;
            list-style:         decimal-leading-zero none inside;
            overflow:           auto; 
        }
        
        #$id ol.codeExcerpt li {
            border-bottom:      1px solid #ddd;
            padding:            2px;
        }
        
        #$id ol.codeExcerpt li.current-line {
            background-color:   #dfd;
            cursor:             pointer;
            font-weight:        bold;
        }
        
        #$id ol.codeExcerpt pre {
            display:            inline;
        }
        
        </style>";
    }
    
    /**
     * Returns the JavaScript code for the debug container.
     * 
     * @param  string $id Identifier for the container
     * @return string
     */
    protected function _getJsCode($id)
    {
        return "<script type='text/javascript'>
        new Draggable('$id', {'handle': 'handle', 'starteffect': '', 'endeffect': ''});
        Event.observe(window, 'load', function() {
            if ($('emzee-dev-debug-main-toggle') != undefined) {
                Event.observe('emzee-dev-debug-main-toggle', 'click', function(event) {
                    $('emzee-dev-debug-container-content').toggle();
                });
            }
            if ($('hierarchy-header') != undefined) {
                Event.observe('hierarchy-header', 'click', function(event) {
                    $('hierarchy-list').toggle();
                });
            }
            if ($('methods-header') != undefined) {
                Event.observe('methods-header', 'click', function(event) {
                    $('methods-list').toggle();
                });
            }
            if ($('children-header') != undefined) {
                Event.observe('children-header', 'click', function(event) {
                    $('children-list').toggle();
                });
            }                
            if ($('backtrace-header') != undefined) {
                Event.observe('backtrace-header', 'click', function(event) {
                    $('backtrace-list').toggle();
                });
            }
            
        });
        
        </script>";
    }
    
    /**
     * Normalises the string (transforms special characters etc.).
     *
     * @param  string $string
     * @return string
     */
    protected function _normalise($string)
    {
        $searchReplace = array(
          '.' => '-',
          '_' => '-',
          ' ' => '-',
          'ä' => 'ae',
          'ö' => 'oe',
          'ü' => 'ue',
          'ß' => 'sz'
        );
        
        $result = strtolower((string)$string);
        $result = str_replace(array_keys($searchReplace), array_values($searchReplace), $result);
        
        return $result;
    }
    
    /**
     * Returns an unique HTML id.
     * 
     * @return string
     */
    public function getHtmlId()
    {
        return 'debug-info-' . $this->_normalise(microtime());
    }
}


app/code/community/Emzee/Dev/Model/SymlinkFix.php
<?php

class Emzee_Dev_Model_SymlinkFix
{
    /**
     * Original value of the 'allow symlinks' setting.
     * 
     * @var string|false
     */
    protected $_originalValue = null;
    
    /**
     * Resets the 'allow symlinks' setting for templates after debug HTML has
     * been generated.
     * 
     * @return void
     */
    public function disableSymlinksForTemplates()
    {
        if (Mage::app()->getUpdateMode() === true) {
            Mage::getConfig()->setNode($this->getAllowSymlinkPath(), $this->_originalValue);
        }
    }
    
    /**
     * Tells Magento to allow symlinks for templates if Magento is in upgrade
     * mode (where the database settings are not loaded yet).
     * 
     * @return void
     */
    public function enableSymlinksForTemplates()
    {
        if (Mage::app()->getUpdateMode() === true) {
            $this->_originalValue = Mage::getConfig()->getNode($this->getAllowSymlinkPath());
            Mage::getConfig()->setNode($this->getAllowSymlinkPath(), 1);
        }
    }
    
    /**
     * Returns the xpath for the default store 'allow symlinks' setting.
     * 
     * @return string|false
     */
    public function getAllowSymlinkPath()
    {
        return 'stores/' . Mage_Core_Model_App::DISTRO_STORE_CODE . '/' . Mage_Core_Block_Template::XML_PATH_TEMPLATE_ALLOW_SYMLINK;
    }    
}


app/code/community/Emzee/Dev/Model/Debug/Object.php
<?php

class Emzee_Dev_Model_Debug_Object
{
    protected $_publicMethods = null;
    
    /**
     * @var ReflectionClass
     */
    protected $_reflectionClass = null;
    
    /**
     * @var mixed
     */
    protected $_variable = null;
    
    /**
     * Calculates the hex color for the class. We fade from
     * red to green, so we have a maximum of 512 shades to play with
     * (from 255,0,0 to 255,255,0 and from 255,255,0 to 0,255,0).
     * 
     * @param  int $shades
     * @param  int $index
     * @return string
     */
    protected function _getColorHex($shades, $index)
    {
        $shades = (int)$shades;
        $index  = (int)$index;
        
        if ($index > $shades) {
            throw new Exception(Mage::helper('emzee_dev')->__("Index must not be higher than count of shades."));
        }
        
        // special case: with one shade, we return the value for green immediately
        if ($shades == 1) {
            return '#00FF00';
        }
        
        $part = 512 / ($shades - 1);
        
        $value = round($part * $index);
        
        /**
         * @refactor: workaround: value 256 has to be 255 actually.
         */
        if ($value == 256) {
            $value = 255;
        }
        
        if ($value <= 255) {
            return '#FF' . ($value < 16 ? '0' : '') . dechex($value) . '00';
        } else {
            $value = 512 - $value;
            return '#' . ($value < 16 ? '0' : '') . dechex($value) . 'FF00';
        }
    }
    
    /**
     * @var mixed $variable
     */
    public function __construct($variable)
    {
        $this->_variable = $variable;
    }
    
    /**
     * Returns the class hierarchy.
     * 
     * @return array
     */
    public function getClassHierarchy()
    {
        $class = new ReflectionClass($this->_variable);
        
        $hierarchy = array(array(
            'file' => $class->getFileName(),
            'name' => $class->getName()
        ));
        
        while ($parent = $class->getParentClass()) {
            array_unshift($hierarchy, array(
                'file' => $parent->getFileName(),
                'name' => $parent->getName()
            ));
            $class = $parent;
        }
        
        $count = count($hierarchy);
        $i = 0;
        foreach ($hierarchy as &$level) {
            $level['color'] = $this->_getColorHex($count, $i++);
        }
        
        return $hierarchy;
    }
    
    /**
     * @return array
     */
    public function getPublicMethods()
    {
        if ($this->_publicMethods === null) {
            $methods = array();
            foreach ($this->getReflectionClass()->getMethods(ReflectionMethod::IS_PUBLIC) as $method) {
                $methods[$method->getName()] = $method;
            }
            ksort($methods);
            $this->_publicMethods = $methods;
        }
        
        return $this->_publicMethods;
    }
    
    /**
     * Returns a reflection of the class.
     * 
     * @return ReflectionClass
     */
    public function getReflectionClass()
    {
        if (is_null($this->_reflectionClass)) {
            $this->_reflectionClass = new ReflectionClass($this->_variable); 
        }
        return $this->_reflectionClass;
    }
    
    /**
     * Returns the original variable.
     * 
     * @return mixed
     */
    public function getVariable()
    {
        return $this->_variable;
    }
    
    /*
     * Returns whether the variable has public methods.
     * 
     * @return boolean
     */
    public function hasPublicMethods()
    {
        return (count($this->getPublicMethods()) > 0);
    }
    
    public function isBlockClass()
    {
        return $this->_variable instanceof Mage_Core_Block_Abstract;
    }
}


app/code/community/Emzee/Dev/Block/Info/Backtrace.php
<?php

class Emzee_Dev_Block_Info_Backtrace extends Mage_Core_Block_Template
{
    /**
     * Number of rows before and after the line in question to be displayed.
     * 
     * @var int
     */
    protected $_excerptRows = 3;
    
    /**
     * Generates an excerpt of the code in the file.
     * 
     * @param  string $path  Path to the file (absolute path preferred)
     * @param  int    $line  Main code line number
     * @return string
     */
    protected function _getExcerptHtml($file, $line)
    {
        if ($file == '') {
            return '<p>' . $this->__('File not exists') .'</p>';
        }
         
        if ($line < 1) {
            return '';
        }
        
        // SplFileObject::seek() is zero-based, so we subtract 1.
        $from = $line - $this->_excerptRows - 1;
        if ($from < 0) {
            $from = 0;
        }
        $main = $line - 1;
        $to   = $line + $this->_excerptRows - 1;
        
        $result = "<ol class='codeExcerpt' start='" . ($from + 1)  . "'>\n";
        $file = new SplFileObject($file);
        
        $file->seek($from);
        do {
            if ($file->key() == $main) {
                $result .= "<li class='current-line'><pre>" . htmlentities($file->current()) . "</pre></li>\n";
            } else {
                $result .= "<li><pre>" . htmlentities($file->current()) . "</pre></li>\n";
            }
            $file->next();
        } while ($file->key() <= $to);
        
        $result .= "</ol>";
        
        return $result;
    }
} 

app/code/community/Emzee/Dev/Model/Debug.php
<?php
/**
 * @copyright (c) 2010 Matthias Zeis (admin@matthias-zeis.com)
 * @license   OSL - Open Software Licence 3.0 (http://opensource.org/licenses/osl-3.0.php)
 */
class Emzee_Dev_Model_Debug
{
    /**
     * Prints information that can be gathered from the Magento configuration.
     */
    public function configInfo()
    {
        $html = '';
        
        $config = Mage::getConfig();
        $options = $config->getOptions();
        
        $html .= '<h2>' . Mage::helper('emzee_dev')->__('Directories') . '</h2>';
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Base directory') . ': ' . $options->getBaseDir() . '</p>'; 
        $html .= '<p>' . Mage::helper('emzee_dev')->__('App directory') . ': ' . $options->getAppDir() . '</p>'; 
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Code directory') . ': ' . $options->getCodeDir() . '</p>'; 
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Design directory') . ': ' . $options->getDesignDir() . '</p>'; 
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Configuration directory') . ': ' . $options->getEtcDir() . '</p>'; 
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Library directory') . ': ' . $options->getLibDir() . '</p>';
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Locale directory') . ': ' . $options->getLocaleDir() . '</p>';
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Media directory') . ': ' . $options->getMediaDir() . '</p>';
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Skin directory') . ': ' . $options->getSkinDir() . '</p>';
        $html .= '<p>' . Mage::helper('emzee_dev')->__('Temporary system directory') . ': ' . $options->getSysTmpDir(). '</p>';
        
        $html .= '<h2>' . Mage::helper('emzee_dev')->__('Modules') . '</h2>';
        $html .= '<table>
                    <thead>
                       <tr>
                          <th>' . Mage::helper('emzee_dev')->__('Module') . '</th>
                          <th>' . Mage::helper('emzee_dev')->__('Status') . '</th>
                          <th>' . Mage::helper('emzee_dev')->__('Code pool') . '</th>
                          <th>' . Mage::helper('emzee_dev')->__('Dependencies') . '</th>
                          <th>' . Mage::helper('emzee_dev')->__('Version') . '</th>
                       </tr>
                    </thead>
                    <tbody>';
        $modules = $config->getModuleConfig();
        
        $moduleStatusClassActive = "active";
        $moduleStatusClassInactive = "inactive";
        foreach ($modules->asCanonicalArray() as $moduleName => $information) {
            $html .= "<tr class='" . ($information['active'] == 'true' ? $moduleStatusClassActive : $moduleStatusClassInactive) . "'>";
            $html .= "<td>$moduleName</td>
                      <td>" . ($information['active'] == 'true' ? Mage::helper('emzee_dev')->__('active') : Mage::helper('emzee_dev')->__('inactive'))  .  "</td>
                      <td class='{$information['codePool']}'>{$information['codePool']}</td>
                      <td>" . ((array_key_exists('depends', $information) && is_array($information['depends'])) ? implode(', ', array_keys($information['depends'])) : '&nbsp;') . "</td>
                      <td>{$information['version']}</td>";
            $html .= "</tr>";
        }
        
        $html .= "</tbody></table>";
        
        /**
         * Complete XML as array
         * 
         * Array-Keys:
         *    [0] => global
         *    [1] => default
         *    [2] => admin
         *    [3] => modules
         *    [4] => frontend
         *    [5] => adminhtml
         *    [6] => install
         *    [7] => stores
         *    [8] => websites
         *    [9] => crontab
         *    [10] => find_feed_attributes
         *    [11] => phoenix
         * 
         * @var array
         */
        $configArray = $config->getNode()->asArray();
        
        /**
         * Entries within area "global":
         *    [0] => install
         *    [1] => resources
         *    [2] => resource
         *    [3] => models
         *    [4] => crypt
         *    [5] => blocks
         *    [6] => helpers
         *    [7] => template
         *    [8] => cache
         *    [9] => session
         *    [10] => request
         *    [11] => log
         *    [12] => eav_frontendclasses
         *    [13] => eav_attributes
         *    [14] => page
         *    [15] => events
         *    [16] => currency
         *    [17] => cms
         *    [18] => index
         *    [19] => fieldsets
         *    [20] => customer
         *    [21] => catalog
         *    [22] => catalogrule
         *    [23] => catalogindex
         *    [24] => payment
         *    [25] => sales
         *    [26] => pdf
         *    [27] => ignoredModules
         *    [28] => ignore_user_agents
         *    [29] => widget
         *    [30] => wishlist
         *    [31] => importexport
         *    [32] => external_cache
         *    [33] => newsletter
         *    [34] => disable_local_modules
         *    [35] => session_save
         */
        
        /**
         * Entries within area "frontend":
         *    [0] => routers
         *    [1] => translate
         *    [2] => layout
         *    [3] => secure_url
         *    [4] => events
         *    [5] => default
         *    [6] => catalog
         *    [7] => product
         *    [8] => category
         *    [9] => cache
         */
        
        echo $html;
    }
    
    /**
     * Returns information on the provided variable.
     * 
     * @param  mixed  $variable
     * @param  string $label    Bezeichnung
     * @return string
     */
    public function info($variable, $label = '')
    {
        if (!Mage::helper('core')->isDevAllowed()) {
            return '';
        }

        switch (gettype($variable)) {
            case 'object':
                $template = 'emzee_dev/info/type/object.phtml';
                $variable      = Mage::getModel('emzee_dev/debug_object', $variable);
                break;
            default:
                $template = 'emzee_dev/info/type/default.phtml'; 
        }
        
        $symlinkFix = Mage::getModel('emzee_dev/symlinkFix');
        $symlinkFix->enableSymlinksForTemplates();
        
        $block = Mage::app()->getLayout()->createBlock('emzee_dev/info')
            ->setTemplate($template)
            ->assign('label', $label)
            ->assign('variable', $variable);
                                         
        $classHierarchy = Mage::app()->getLayout()->createBlock('emzee_dev/info_classHierarchy', 'classHierarchy')
            ->setNameInLayout('classHierarchy')
            ->setTemplate('emzee_dev/info/classHierarchy.phtml')
            ->assign('variable', $variable);
        
        $blockClass = Mage::app()->getLayout()->createBlock('core/template', 'blockClass')
            ->setNameInLayout('blockClass')
            ->setTemplate('emzee_dev/info/blockClass.phtml')
            ->assign('variable', $variable);
        
        $backtrace = Mage::app()->getLayout()->createBlock('emzee_dev/info_backtrace', 'backtrace')
            ->setNameInLayout('backtrace')
            ->setTemplate('emzee_dev/info/backtrace.phtml')
            ->assign('backtrace', debug_backtrace());
                                             
        $block->append($classHierarchy, 'classHierarchy');
        $block->append($blockClass, 'blockClass');
        $block->append($backtrace, 'backtrace');
        
        $html = $block->toHtml();
        
        $symlinkFix->disableSymlinksForTemplates();
        
        return $html;                               
    }
    
    /**
     * Returns the amount of allocated RAM in this very moment.
     *
     * @param  boolean $realUsage The same parameter as for memory_get_usage()
     * @param  boolean $rawData Returns the size without "beautified" size (and as integer) 
     * @return string|int
     */
    public function memoryUsage($realUsage = false, $rawData = false)
    {
        if (!Mage::helper('core')->isDevAllowed()) {
            return '';
        }
        
        $size = memory_get_usage((bool)$realUsage);
        
        if ((bool)$rawData) {
            return $size;
        } 
        
        $unit = array('b','kb','mb','gb','tb','pb');
        return @round($size/pow(1024,($i=floor(log($size,1024)))),2).' '.$unit[$i];
    }
    
    public function modelInfo()
    {
        if (!Mage::helper('core')->isDevAllowed()) {
            return '';
        }
        
        $config = Mage::getConfig()->getNode()->global->models->asArray();
        ksort($config);
        
        $groups = array_keys($config);
        
        $html = '
        <style type="text/css">
        #emzee_dev_modelinfo {
            background-color: #fff;
            border: 1px solid #ccc;
            padding:    10px;
            text-align: left;
        }
        
        #emzee_dev_modelinfo li {
            border-bottom:  1px solid #eef;
            margin: 0 0 10px;
            padding: 0 0 5px;
        }
        
        #emzee_dev_modelinfo .error {
            color:  #f00;
            font-weight:    bold;
        }
        </style> 
        
        <ol id="emzee_dev_modelinfo">';
        
        foreach ($config as $group => &$groupData) {
            $html .= '<li id="group-' . $group . '">';
            $html .= "<strong>$group</strong>:<br>" . Mage::helper('emzee_dev')->__('To be found in') . ": " . $groupData['class'];
            unset($groupData['class']); // @todo delete: only used for finding all informations to show
            
            if (array_key_exists('resourceModel', $groupData)) {
                $html .= "<br>" . Mage::helper('emzee_dev')->__('Resource model') . ": ";
                if (in_array($groupData['resourceModel'], $groups)) {
                    $html .= "<a href='#group-{$groupData['resourceModel']}'>{$groupData['resourceModel']}</a>";
                } else {
                    $html .= "<span class='error'>" . Mage::helper('emzee_dev')->__("Error: Group '%s' expected, but not defined", $groupData['resourceModel']) . "</span>";
                }
                unset($groupData['resourceModel']); // @todo delete: only used for finding all informations to show
            }
            
            if (array_key_exists('entities', $groupData)) {
                $html .= "<br>" . Mage::helper('emzee_dev')->__('Entities') . ": ";
                $entities = array();
                foreach ($groupData['entities'] as $entityName => &$entityData) {
                    if (array_key_exists('table', $entityData)) {
                        $entities[] = "<span title='{$entityData['table']}'>$entityName</span>";
                        unset($entityData['table']); // @todo delete: only used for finding all informations to show
                    }
                    
                    if (empty($entityData)) {
                        unset($groupData['entities'][$entityName]); // @todo delete: only used for finding all informations to show
                    }
                    
                    if (empty($groupData['entities'])) {
                        unset($groupData['entities']); // @todo delete: only used for finding all informations to show
                    }
                }
                $html .= implode(', ', $entities);
            }
            
            $html .= '</li>';
            
            if (empty($groupData)) {
                unset($config[$group]);  // @todo delete: only used for finding all informations to show
            }
        }
        
        $html .= '</ol>';
        
        return $html;
        
        /*
        if (!empty($config)) {
            $this->info($config, '', true, '', 'text-align:left;');  // @todo delete: only used for finding all informations to show
        }
        */
        
    }
}
